find_similar.py

```diff
@@ -0,0 +1,155 @@
+import torch
+import pyodbc
+from scipy.spatial.distance import cdist, cosine
+import numpy as np
+import json
+import chess
+import re
+
+import settings
+import model
+from inference import Inference
+
+
+def key(i):
+    return i[1]
+
+def l1_loss(matrix,target):
+    return [np.sum(np.abs(np.array(m)-np.array(target))) for m in matrix]
+    
+def nearest(matrix,target):
+    return cdist(matrix, np.atleast_2d([target]))
+    
+def cosine_dist(matrix,target):
+    return [cosine(m, target) for m in matrix]
+    
+similarity_functions = {"Closest vectors":nearest,"L1 Loos":l1_loss,"Cosine distance":cosine_dist}
+
+def get_move(line):
+    line = str(line)
+    game = []  # list of moves in pgn notation
+    for move in re.findall(
+        r"[^{\[.}\]]+ ", line.replace("?", "").replace("!", "")
+    ):  # Extract all moves without comments
+        game.extend(
+            move.strip().split(" ")
+        )  # Sometimes one string contains two moves with space between
+    return game
+
+
+def find_lowest(array, num, key=key):
+    """
+    Finds x smallest values ​​in a table.
+    :param array: list of records
+    :param num: number of records to be found
+    :param key: sort function key
+    :return: list of index and records
+    """
+    result = []
+    for i, a in enumerate(array):
+        if len(result) < num:
+            result.append((i, a))
+            result.sort(key=key)

```

Comment:
why not just one sort and then returning first num values?

Reply:
because sorting the whole is less efficient, even with a small dataset